File: 146.lru_cache.py
"""
Design a data structure that follows the constraints of a Least Recently Used
(LRU) cache.

Implement the LRUCache class:

LRUCache(int capacity) Initialize the LRU cache with positive size capacity.
int get(int key) Return the value of the key if the key exists, otherwise
return -1.  void put(int key, int value) Update the value of the key if the key
exists. Otherwise, add the key-value pair to the cache. If the number of keys
exceeds the capacity from this operation, evict the least recently used key.
The functions get and put must each run in O(1) average time complexity.

"""

class DLNode:
    def __init__(self, pre=None, nxt=None, key=None, val=None):
        self.pre = pre
        self.nxt = nxt
        self.key = key
        self.val = val
        
    def insert(self, pre, nxt):
        self.pre = pre
        self.nxt = nxt
        pre.nxt = self
        nxt.pre = self
    
    def excise(self):
        self.pre.nxt = self.nxt
        self.nxt.pre = self.pre


class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.old = DLNode()
        self.new = DLNode()
        self.old.nxt = self.new
        self.new.pre = self.old
        self.ent = dict()
        
    def _touch(self, node):
        node.excise()
        node.insert(self.new.pre, self.new)
        
        
    def get(self, key: int) -> int:
        if key not in self.ent:
            return -1
        self._touch(self.ent[key])
        return self.ent[key].val
        
    def put(self, key: int, value: int) -> None:
        if key in self.ent:
            nd = self.ent[key]
            nd.val = value
            self._touch(nd)
        else:
            if len(self.ent) == self.capacity:
                del self.ent[self.old.nxt.key]
                self.old.nxt.excise()
            
            nd = DLNode(None, None, key, value)
            nd.insert(self.new.pre, self.new)
            self.ent[key] = nd
            
            
                
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146.lru_cache.cc
class DLNode {
  // Doubly-linked list node
  public:
  int key, val;
  DLNode *pre, *nxt; // non-ownership relation

  DLNode(int key, int val, DLNode *pre=nullptr, DLNode *nxt=nullptr) 
    : key(key), val(val), pre(pre), nxt(nxt) { }

  void excise() {
    // excise oneself from the list
    pre->nxt = nxt;
    nxt->pre = pre;
    pre = nullptr;
    nxt = nullptr;
  }

  void insert_after(DLNode *left) {
    DLNode *right = left->nxt;
    left->nxt = this;
    assert(right);
    right->pre = this;
    pre = left;
    nxt = right;
  }

};

class LRUCache {
  public:
  shared_ptr<DLNode> head, tail;
  int capacity;
  map<int, shared_ptr<DLNode>> nodes;

public:
    LRUCache(int capacity) : capacity(capacity) {
      head = make_shared<DLNode>(-1, -1);
      tail = make_shared<DLNode>(-1, -1);
      head->nxt = tail.get();
      tail->pre = head.get();
    }

    
    int get(int key) {
      if (!nodes.contains(key)) return -1;
      auto node = nodes[key];
      node->excise();
      node->insert_after(head.get());
      return node->val;
    }
    
    void put(int key, int value) {
      if (!nodes.contains(key)) {
        // cout << "did not find " << key << endl;
        if (nodes.size() == capacity) {
          auto lru = tail->pre;
          lru->excise();
          // cout << "full, deleting " << lru->key << endl;
          nodes.erase(lru->key);
        }
        auto node = std::make_shared<DLNode>(key, value);
        node->insert_after(head.get());
        nodes[key] = node;
      } else {
        auto node = nodes[key];
        node->excise();
        node->insert_after(head.get());
        node->val = value;
      }
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

 /*
I assume both get and put bump the key to most recently used position.

So, you need to keep an ordering (doubly linked list?)


 */
